File: java/src/main/java/com/jackson/chapter1/Problem1_3.java
package com.jackson.chapter1;

public class Problem1_3 {

    /**
     * Kadane's solution - computes in one pass
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public static int maxSubarraySum(int[] array) {
        int maxEndingHere = 0;
        int maxSoFar = 0;

        for (int i = 0; i < array.length; i++) {
            maxEndingHere = Math.max(array[i], maxEndingHere + array[i]);
            maxSoFar = Math.max(maxEndingHere, maxSoFar);
        }

        return maxSoFar;
    }
}

File: java/src/main/java/com/jackson/chapter1/Problem1_3.kt
package com.jackson.chapter1

import kotlin.math.max

/**
 * Kadane's solution - computes in one pass
 * Time complexity: O(n)
 * Space complexity: O(1)
 */
fun maxSubarraySum(array: IntArray): Int {
    var maxEndingHere = 0
    var maxSoFar = 0

    array.forEach {
        maxEndingHere = max(it, maxEndingHere + it)
        maxSoFar = max(maxEndingHere, maxSoFar)
    }
    return maxSoFar
}


/**
 * Bruteforce solution that uses nested loops
 * Time complexity: O(n²)
 * Space complexity: O(1)
 */
fun maxSubarraySumBruteForce(array: IntArray): Int {
    var maxSum = Integer.MIN_VALUE
    for (i in array.indices) {
        var subarraySum = 0
        for (k in i until array.size) {
            subarraySum += array[k]
        }
        if (subarraySum > maxSum) {
            maxSum = subarraySum
        }
    }
    return maxSum
}
